File: src/main/java/frc/team449/control/DriveSubsystem.java
package frc.team449.control;

import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.wpilibj2.command.Subsystem;
import io.github.oblarg.oblog.Loggable;

/** A drivetrain that uses closed-loop velocity control. */
public interface DriveSubsystem extends Subsystem, Loggable {
  /** Set the desired speeds to go at. */
  void set(ChassisSpeeds desiredSpeeds);

  Rotation2d getHeading();

  /** Reset the drivetrain's pose. Should really only be done at the start of auto. */
  void setPose(Pose2d pose);

  Pose2d getPose();

  /** Set both motors' voltage to 0. */
  void stop();
}

File: src/main/kotlin/frc/team449/control/DriveSubsystem.kt
package frc.team449.control

import edu.wpi.first.math.MathUtil
import edu.wpi.first.math.geometry.Pose2d
import edu.wpi.first.math.geometry.Rotation2d
import edu.wpi.first.math.geometry.Translation2d
import edu.wpi.first.math.kinematics.ChassisSpeeds
import edu.wpi.first.wpilibj2.command.Subsystem
import io.github.oblarg.oblog.Loggable
import kotlin.math.PI

/** A drivetrain that uses closed-loop velocity control. */
interface DriveSubsystem : Subsystem, Loggable {
  var heading: Rotation2d
    get() = Rotation2d(MathUtil.inputModulus(this.pose.rotation.radians, -PI, PI))
    set(value) {
      this.pose = Pose2d(Translation2d(this.pose.x, this.pose.y), value)
    }

  var pose: Pose2d

  /** Set the desired speeds to go at. */
  fun set(desiredSpeeds: ChassisSpeeds)

  /** Set both motors' voltage to 0. */
  fun stop()

  /**
   * Used to simulate a drivetrain. Only one instance of this class should be made per drivetrain.
   */
  interface SimController {
    fun update()

    /**
     * Simulate the current drawn by the drivetrain
     */
    fun getCurrentDraw(): Double
  }
}
